roofit/roofitcore/src/RooRealSumFunc.cxx

```diff
@@ -149,10 +149,10 @@ RooRealSumFunc::RooRealSumFunc(const char *name, const char *title, const RooArg
       _coefList.add(*coef);
    }
 
-   func = (RooAbsReal *)funcIter->Next();
+   func = (RooAbsArg *)funcIter->Next();
```

Comment:
It's not your fault, but you opened Pandora's box 🙂
- No c-style casts in any code that gets touched after 2018
- `assert(0)` is useless in production code. Please throw `std::invalid_argument`.
- Don't use legacy iterators and don't define variables outside of the loop.
- Instead, use something like this:
```c++
for (unsigned int i = 0; i < inFuncList.size(); ++i) {
  auto func = static_cast<RooAbsArg*>(inFuncList[i]);
  auto coef = static_cast<RooAbsArg*>(inCoefList[i]);
  ...
```
or directly `dynamic_cast` to the required types. If one of those doesn't exist or has the wrong type, issue an error message and throw.

Reply:
Fixing coding style is a bit out of scope for a proposed minor compiler warning fix.
There are over 50 assert(0) in the roofit/ source tree, and if they should be changed they should be changed consistently by the code maintainers. Changing the in just one file that happened to generate a compiler warning would make things inconsistent. Changing assert(0) to an exception throw changes the behaviour of the code drastically and would make the PR to something very different from a simple compiler warning fix. The use of assert could in fact be intentional by the original code writer if the intention is that only a debug build should terminate on these errors.

Reply:
Ok, point taken. Converting this to an issue. 🙂

Reply:
#8374